plumbing/protocol/packp/advcaps.go

```diff
@@ -0,0 +1,95 @@
+package packp
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/go-git/go-git/v5/plumbing/format/pktline"
+	"github.com/go-git/go-git/v5/plumbing/protocol/packp/capability"
+)
+
+// AdvCaps values represent the information transmitted on an the first v2
+// message.  Values from this type are not zero-value
+// safe, use the New function instead.
+type AdvCaps struct {
+	// Service represents the requested service.
+	Service string
+	// Capabilities are the capabilities.
+	Capabilities *capability.List
+}
+
+// NewAdvRefs returns a pointer to a new AdvRefs value, ready to be used.
+func NewAdvCaps() *AdvCaps {
+	return &AdvCaps{
+		Capabilities: capability.NewList(),
+	}
+}
+
+// IsEmpty returns true if doesn't contain any capability.
+func (a *AdvCaps) IsEmpty() bool {
+	return a.Capabilities.IsEmpty()
+}
+
+func (a *AdvCaps) Encode(w io.Writer) error {
+	pe := pktline.NewEncoder(w)
+	pe.EncodeString("# service=" + a.Service + "\n")
+	pe.Flush()
+	pe.EncodeString("version 2\n")
+
+	for _, c := range a.Capabilities.All() {
+		vals := a.Capabilities.Get(c)
+		if len(vals) > 0 {
+			pe.EncodeString(c.String() + "=" + strings.Join(vals, " ") + "\n")
+		} else {
+			pe.EncodeString(c.String() + "\n")
+		}
```

Comment:
Not sure if this suggestion works, since each invocation of `EncodeString` will create the line size prefix (part of the pktline protocol). And on v2 all that payload must be on the same line.

Reply:
You are absolutely right, updated it to handle that as a var of its own.